Add EIP-5646: Token state fingerprint (#5646)
ethereum/EIPs: EIPS/eip-5646.md

---
eip: 5646

> title: Token state fingerprint

Reviewer: ```suggestion
title: Token State Fingerprint
```

> description: Define a way how to distinguish between different states of the same token without implementation details.

Reviewer: This is a bit long-winded. I recommend:
```suggestion
description: Opaque states for EIP-721 NFTs
```

author: Naim Ashhab (@ashhanai)
discussions-to: https://ethereum-magicians.org/t/eip-5646-discussion-token-state-fingerprint/10808
status: Draft
type: Standards Track
category: ERC
created: 2022-09-11
requires: 165, 721
---

## Abstract

> Extend [EIP-721](./eip-721.md) standard with new function `getStateFingerprint`, which returns a unique 32 bytes fingerprint of a token state. This enables distinguishing between different states of the same token and referencing a specific token state without knowing anything about its state structure. It does not define *how* should be state fingerprint computed or stored.

Reviewer: ```suggestion
This EIP extends [EIP-721](./eip-721.md) with a function, `getStateFingerprint`, which returns a unique 32-byte fingerprint of the state of a given token ID. This enables applications to distinguish between different states of the same token and reference a specific token state without knowing anything about its details. It does not define *how* the state fingerprint should be computed.
```

## Motivation

> More and more non-fungible tokens are starting to be "dynamic" in a sense, that the owner can change tokens state, which could affect tokens intrinsic value. A lot of marketplaces have to start taking this into account when managing users offers to buy or accept a particular token as a collateral.
> 
> Currently there is no UX friendly way to prevent cases, where a lender makes an offer to a dynamic NFT and borrower withdraws most of its value before accepting the offer. One solution is to use wrapper / bundler, to disable owners' right to modify tokens state, but this brings additional inconveniences. It costs unnecessary gas and users cannot use tokens utility, that doesn't change tokens state (voting rights, NFT gate access, airdrops).
> 
> Instead of using wrappers, marketplaces can use state fingerprints as a part of an offer. In that case, an offer that was made to a token in one state, would be invalid for the same token (same token id) in a different state.
> 
> Some attempts to standardize token state already existed in the past, for example [Non-Fungible Token State Verification](https://github.com/ethereum/EIPs/issues/1180) and some tokens already implemented similar functions in their contracts, see [MokensDelegate](https://etherscan.io/address/0x8e332fe4df2876cce3b9e6cab160c61ef4c31f77#code)s function `getStateHash(uint256 _tokenId) public view returns (bytes32 stateHash)`. Unfortunately, no EIP standard of this kind has been adopted to this day.

Reviewer: ```suggestion
An increasing number of NFTs are becoming "dynamic," where owners can change tokens' states. A lot of marketplaces have begun taking this into account when managing users' offers to, for example, buy or accept a particular token as collateral.

Currently, there is no application-friendly way to prevent cases where a lender makes an offer to a dynamic NFT and the borrower expends most of the NFT's value before accepting the offer. One solution is to use a wrapper/bundler in order to disable owners' right to modify tokens' states, but this is inconvenient, costing unnecessary gas and preventing the use of tokens in ways that don't change tokens' states (voting rights, NFT gate access, airdrops).

Instead of using wrappers, marketplaces can use state fingerprints as a part of an offer. In that case, an offer that was made to a token in one state would be invalid for the same token in a different state.
```
No external links.

###  Type of dynamic tokens
- tokenized LP positions (Uniswap v3)
- gaming assets (CryptoKitties)
- dynamic wrappers / bundlers

### Use cases
- NFT marketplace orders / offers
- on-chain metadata uri updates

> - front running

Reviewer: Would you mind elaborating? This is currently unclear.

## Specification
The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

> The standard defines one view function.

Reviewer: ```suggestion
```
This is clear from the context.

```solidity
pragma solidity ^0.8.0;

> interface ERC5646 is IERC721 {

Reviewer: ```suggestion
interface ERC5646 is ERC721 {
```
The interface defined in EIP-721 is called `ERC721` not `IERC721`.

	/// @notice Function to return current token state fingerprint.
	/// @param tokenId Id of a token state in question.
	/// @return Current token state fingerprint.
	function getStateFingerprint(uint256 tokenId) external view returns (bytes32);
}
```

> This standard doesn't define *how* should be the `getStateFingerprint` function implemented.
> 

Reviewer: ```suggestion
```
Implicit

> State fingerprint MUST be different for a different token state.
> State fingerprint MUST be the same for the same token state.
> State fingerprint MUST include all parameters that change the intrinsic value of a token.
> Two state fingerprints of two different token contracts MUST NOT be compared between each other.
> State fingerprint MUST NOT be used as a sole token identifier and MUST be used with at least token contract address and token id, to identify a token.
> Function `supportsInterface(bytes4)` MUST return `true` for input `0xf5112315`.
> Token uri MAY be included in a state fingerprint, but it's rather encouraged to implement [EIP-4906](./eip-4906.md) for that purpose.

Reviewer: ```suggestion
- `getStateFingerprint` MUST return a different value when the token state changes.
- `getStateFingerprint` MUST NOT return a different value when the token state remains the same.
- `getStateFingerprint` MUST factor in all parameters that might change the intrinsic value of a token.
- `getStateFingerprint(tokenId1)` MAY be different than `getStateFingerprint(tokenId2)` even if they have the same state.
- `getStateFingerprint` MUST NOT be solely used to uniquely identify a token.
- `supportsInterface(0xf5112315)` MUST return `true`.
```

## Rationale
To be as flexible as possible, this standard doesn't define *how* should be the `getStateFingerprint` function implemented. Simple hash of an abi encoded state parameters is expected to be enough as a state fingerprint, though some tokens could implement a more gas efficient way for their specific state. Some implementations could compute the fingerprint on the fly, where another can store it and update its value with every state-mutating call.

It is not necessary to prevent state fingerprint collision between two different token contracts as state fingerprints should be compared in a context of one token contract. Because of that, implementation could be more light-weight than, for example, [EIP-712](./eip-712.md), which has to define domain separator to prevent this case. It is encouraged to use state fingerprints as a part of [EIP-712](./eip-712.md) type struct though.

This standard, contrary to [EIP-1948](./eip-1948.md), doesn't define a function for directly updating state fingerprints. The intention is to update state fingerprints only via state-mutating calls.

> In comparison to [Non-Fungible Token State Verification](https://github.com/ethereum/EIPs/issues/1180), this standard leaves the decision, which state parameters should be included in a state fingerprint, to token developers. This should lead to easier adoption of the standard as marketplaces don't have to compose the state fingerprint themselves.

Reviewer: No external links
